[IOPLT-717] Add listTrial to CosmosDB and http layer

PascalDR opened the pull request:
<!---
Please always add a PR description as if nobody knows anything about the context
these changes come from. Even if we are all from our internal team, we may not
be on the same page. Write this PR as you were contributing to a public OSS
project, where nobody knows you and you have to earn their trust. This will
improve our projects in the long run!

Thanks :).
-->

#### List of Changes
<!--- Describe your changes in detail -->
- Added `list` in CosmosDB trial adapter with related tests
- Added `listTrials` endpoint with related option type and tests
- Added `parseQueryParameter` function

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
These changes allows an `ApiTrialSupport` user to query the API for all the trials

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested locally (connected to dev environment) by running variations of the following curl and verifying the returned data:
```
curl --location 'http://localhost:7071/trials?pageSize=2&maximumId={{maxId}}&minimumId={{minId}}' \
--header 'X-User-Id: asd' \
--header 'X-User-Groups: ApiTrialSupport' \
--header 'Accept: application/json' \
--header 'X-Functions-Key: {{apiKey}}'
```

#### Checklist before requesting a review
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have performed a self-review of my code.
- [ ] I have committed only changes relevant to the task.
- [x] I have minimized the number of changes.
- [ ] My changes are about only one task or issue.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

PascalDR committed:
Implemented getAll trials functionality

PascalDR committed:
Implemented handler

PascalDR committed:
Added route handler

kin0992 commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -51,6 +51,7 @@ export interface TrialWriter {
 
 export interface TrialReader {
   readonly get: (trialId: TrialId) => TE.TaskEither<Error, O.Option<Trial>>;
+  readonly getAll: () => TE.TaskEither<Error, readonly Trial[]>;
```
**suggestion**
```suggestion
  readonly list: () => TE.TaskEither<Error, readonly Trial[]>;
```

I would not use the `getAll` name, because it is highly improbable we will return all the trials

kin0992 reviewed (commented):
i left some comments.

you can see the meaning the prefixes of the comments (like `suggestion`, `nitpick` and so on) checking [the conventional comments specification](https://conventionalcomments.org/)

PascalDR committed:
Merge branch 'main' of https://github.com/pagopa/trial-system into feature/get_trials

PascalDR committed:
Fix user validation function

PascalDR commented on apps/functions-subscription/src/adapters/azure/functions/get-trials.ts:
```diff
@@ -0,0 +1,28 @@
+> = H.of((req: H.HttpRequest) =>
+  pipe(
+    RTE.ask<Env>(),
+    RTE.apFirst(RTE.fromEither(verifyUserGroup(['ApiTrialSupport'])(req))),
```
Resolved in [6eab882](https://github.com/pagopa/trial-system/pull/244/commits/6eab88204536d22a9f8eb0520135ad3e30c33a97)

PascalDR committed:
Update apps/functions-subscription/src/adapters/azure/cosmosdb/trial.ts

Co-authored-by: Marco Comi <9998393+kin0992@users.noreply.github.com>

PascalDR committed:
Update apps/functions-subscription/src/domain/trial.ts

Co-authored-by: Marco Comi <9998393+kin0992@users.noreply.github.com>

PascalDR committed:
fixed method name

PascalDR committed:
removed toTrialAPIArray and refactored code

PascalDR committed:
changed function name

PascalDR commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -88,3 +89,9 @@ export const getTrialById = (trialId: TrialId) =>
     RTE.flatMapTaskEither(({ trialReader }) => trialReader.get(trialId)),
   );
+
+export const getAllTrials = () =>
```
Resolved in [cbc5514](https://github.com/pagopa/trial-system/pull/244/commits/cbc5514c5cad00373e8ada027e762169eb5de49f)

PascalDR commented on apps/functions-subscription/src/system-env.ts:
```diff
@@ -23,6 +23,7 @@ export const makeSystemEnv = (capabilities: Capabilities) => ({
   updateActivationJob: flow(updateActivationJob, apply(capabilities)),
   createTrial: flow(insertTrial, apply(capabilities)),
   getTrial: flow(getTrialById, apply(capabilities)),
+  getTrials: flow(getAllTrials, apply(capabilities)),
```
Resolved in [cbc5514](https://github.com/pagopa/trial-system/pull/244/commits/cbc5514c5cad00373e8ada027e762169eb5de49f)

PascalDR committed:
Moved route handler

PascalDR commented on apps/functions-subscription/src/main.ts:
```diff
@@ -204,6 +205,13 @@ app.http('getTrial', {
   route: 'trials/{trialId}',
 });
 
+app.http('getTrials', {
```
Resolved in [123485e](https://github.com/pagopa/trial-system/pull/244/commits/123485e228436264ba767f17259e16359bef8e60)

robbcocco commented on apps/functions-subscription/src/main.ts:
```diff
@@ -200,6 +201,13 @@ app.http('createTrial', {
   route: 'trials',
 });
 
+app.http('getTrials', {
```
I would change 'getTrials' to 'listTrials' to make the code consistent.

PascalDR committed:
renamed file and handler maker

PascalDR committed:
fixed names

PascalDR committed:
added tests

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-traials.test.ts:
Suggestion: file should be renamed in list-trials.test.ts

robbcocco committed:
Add simple cursor pagination for list trials.

robbcocco committed:
Update pagination parameters for listtrials to be optional.

robbcocco committed:
Update listtrials response to openapi specification.

robbcocco committed:
Updated tests.

robbcocco committed:
Merge remote-tracking branch 'origin/main' into feature/get_trials

kin0992 reviewed (commented):
I dropped some comments, I'm not finished yet, because the PR contains both the HTTP and the CosmosDB layers.

I'm gonna do another review later

PascalDR committed:
Update apps/functions-subscription/src/adapters/azure/functions/__tests__/data.ts

Co-authored-by: Marco Comi <9998393+kin0992@users.noreply.github.com>

PascalDR committed:
Update apps/functions-subscription/src/adapters/azure/cosmosdb/trial.ts

Co-authored-by: Marco Comi <9998393+kin0992@users.noreply.github.com>

robbcocco committed:
Updated listTrials query.

robbcocco committed:
Update list trials test description.

robbcocco committed:
Remove null type for maximumId and minimumId in listTrials.

robbcocco committed:
Update list trial signature to take ListTrialOptions instead of three parameters.

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,88 @@
+import { HttpRequest } from '@azure/functions';
+import { Trial } from '../../../../domain/trial';
+
+describe('makeGetTrialHandler', () => {
```
Corrected on c7edf75e405bae041ef45c40322d2fecb512c294

robbcocco commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -52,6 +52,11 @@ export interface TrialWriter {
+    pageSize: number,
+    maximumId?: TrialId | null,
+    minimumId?: TrialId | null,
+  ) => TE.TaskEither<Error, readonly Trial[]>;
```
Removed the null type and created an object in 457f69c8636c890d252af4ecabd738c81346fb33 and 21d55b92ddaacc4fe909d91d4f73f9256199980e

robbcocco committed:
Update listTrials handler response type.

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,81 @@
+
+const makeHandlerKitHandler: H.Handler<
+  H.HttpRequest,
+  | H.HttpResponse<{
```
Updated on e0622ebbbf46a07aaace981f54601e57ec7b1728

robbcocco committed:
Update listtrials tests

robbcocco committed:
Update listTrial test types

robbcocco commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -22,6 +22,35 @@ describe('insertTrial', () => {
+    const expected = E.right([aTrial1, aTrial2]);
+
+    expect(actual).toMatchObject(expected);
+  });
```
Resolved in a23b03d1487e4a63190813edc8c36170c796f365

robbcocco commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +84,41 @@ describe('makeTrialsCosmosContainer', () => {
+    it('should return list of trials', async () => {
+      const mockDB = makeDatabaseMock();
+
+      const trials = [aTrial1, aTrial2];
```
Resolved in a23b03d1487e4a63190813edc8c36170c796f365

robbcocco commented on apps/functions-subscription/src/domain/__tests__/data.ts:
```diff
@@ -93,3 +93,16 @@ export const aCreatedTrial = {
+  id: 'a2' as TrialId,
+  name: 'n2' as NonEmptyString,
+  description: 'd2',
+};
```
Removed unneccessary objects in a23b03d1487e4a63190813edc8c36170c796f365

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,88 @@
+    const result: { readonly items: readonly Trial[] } =
+      (await actual.json()) as {
+        readonly items: readonly Trial[];
+      };
```
The cast is necessary as actual.json() returns an object of type `unknown`.
In a6d0d721a4b65d9ff7fc69b0af2d84da8ddf8401 I updated the type to `TrialPaginatedCollection` and cleaned the code.

robbcocco committed:
Create grumpy-singers-fry.md

robbcocco committed:
Merge remote-tracking branch 'origin/main' into feature/get_trials

robbcocco committed:
Merge remote-tracking branch 'origin/main' into feature/get_trials

robbcocco committed:
Updated TrialIdCodec type
Updated trialIds in tests to conform to the specification
Updated open api specifications

kin0992 commented on .changeset/grumpy-singers-fry.md:
```diff
@@ -0,0 +1,5 @@
+"functions-subscription": minor
+---
+
+[IOPLT-746] CosmosDB Trials adapter
```
issue: this does not represent entirely the changes

kin0992 commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -15,12 +18,20 @@ interface TrialIdBrand {
 export const TrialIdCodec = t.brand(
-  NonEmptyString,
-  (str): str is t.Branded<NonEmptyString, TrialIdBrand> => str.length > 0,
+  WithinRangeString(26, 27),
+  (str): str is t.Branded<WithinRangeString<26, 27>, TrialIdBrand> =>
+    str.length >= 26 && str.length < 27,
```
**question**: why this change?

it this change necessary? is it necessary on the domain level?

robbcocco replied:
creating the response object on `apps/functions-subscription/src/adapters/azure/functions/codec.ts` requires to cast a trialId (`NonEmptyString`) to nextId/previousId (`WithinRangeString<26, 27>`).
this could be done in three ways:
- by casting the trialId to unknown before casting it to NonEmptyString
- by changing the trialId type to match nextId/previousId
- by removing the minLength and maxLength properties from nextId/previousId

I went for the second options since `WithinRangeString<26, 27>` matches the properties for the trialId type too and it avoids calls to the api with invalid trialIds.
this also required me to update the open api specification and the trialIds in the tests to match the 26 characters requirement.

kin0992 replied:
I get it, but here you are updating a value on the domain and I don't think we need to be that stricter there.

If within the adapter we need to handle a different type, then I'd manipulate the `TrialId` there, so we decouple the domain to the adapters.

robbcocco replied:
Resolved in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -15,12 +18,20 @@ interface TrialIdBrand {
 );
 export type TrialId = t.TypeOf<typeof TrialIdCodec>;
 
+const ListTrialOptionsCodec = t.strict({
+  pageSize: t.number,
+  maximumId: t.union([TrialIdCodec, t.undefined]),
+  minimumId: t.union([TrialIdCodec, t.undefined]),
+});
```
**suggestion**: improve this type

from a developer point of view, this type force to explicit the `maximumId` and the `minimumId`, even if the parameters can be `undefined`.

Make it possible to not explicit those parameters, when not needed

robbcocco replied:
Resolved in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -23,6 +24,30 @@ describe('insertTrial', () => {
+    const anotherTrial = {
+      ...aTrial,
+      id: 'anotherTrialId012345678901' as TrialId,
+      name: 'anotherTrialName' as NonEmptyString,
+      description: 'anotherTrialDescription',
```
**nitpick**: you can omit to update every parameter

robbcocco replied:
Removed description in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -23,6 +24,30 @@ describe('insertTrial', () => {
+
+    const actual = await listTrials({
+      pageSize: 1,
+      maximumId: undefined,
+      minimumId: undefined,
```
**suggestion**: as described in a different comment

kin0992 commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -23,6 +24,30 @@ describe('insertTrial', () => {
+      maximumId: undefined,
+      minimumId: undefined,
+    })(testEnv)();
+    const expected = E.right([aTrial, anotherTrial]);
```
**issue**: `pageSize` is 1, but the array returns 2 elements

robbcocco replied:
Updated pageSize in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -23,6 +24,30 @@ describe('insertTrial', () => {
     expect(testEnv.trialWriter.insert).toBeCalledWith(aTrial);
   });
 
+  it('should return all trials', async () => {
```
**suggestion**: the `it` description doesn't represent what you are testing here

robbcocco replied:
Updated description in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
**issue**: the `list` function has various combination of input.

I'd add unit tests, checking the query as well, for the critical inputs.

robbcocco replied:
Added in 8fdbe8398d5cfcd6e97b116f885ad7936c80320f

kin0992 commented on apps/functions-subscription/src/domain/__tests__/data.ts:
```diff
@@ -12,7 +12,7 @@ const aSubscriptionHistoryIdV1 =
 const aSubscriptionId = 'aSubscriptionId' as SubscriptionId;
 const aUserId = 'aUserId' as UserId;
-const aTrialId = 'aTrialId' as TrialId;
+const aTrialId = 'aTrialId012345678901234567' as TrialId;
```
**question**: why is this necessary?

are you sure we need to update this?
IMHO this is not needed in the domain level

robbcocco replied:
Reverted in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/middleware.ts:
```diff
@@ -51,7 +63,7 @@ export const parseHeaderParameter =
     );
 
-type AllowedGroup = 'ApiTrialUser' | 'ApiTrialManager';
+type AllowedGroup = 'ApiTrialUser' | 'ApiTrialManager' | 'ApiTrialSupport';
```
**issue**: this change is in a [different PR](https://github.com/pagopa/trial-system/pull/256/files#diff-54d80969596ade2ef3206d66ed37285303a34f0ab736458d9b9e7089c5166de2R54)

robbcocco replied:
Resolved in 830a067a8470df94ce7f2ed67c6bc641097ad847

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/middleware.ts:
```diff
@@ -34,6 +34,18 @@ export const parsePathParameter =
+/**
+ * Parses a specific query parameter of an HTTP request using the provided schema.
+ */
+export const parseQueryParameter =
```
**suggestion**: add unit test for this function

robbcocco replied:
Added tests in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,94 @@
+    );
+
+    expect(actual.status).toStrictEqual(200);
+    expect(await actual.json()).toMatchObject({ items: [] });
```
**suggestion**
```suggestion
    expect(await actual.json()).toStrictEqual({ items: [] });
```

I think we can use `toStrictEqual`, since there is no need to specify `nextId` and `prevId`

robbcocco replied:
Resolved in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,94 @@
+      makeFunctionContext(),
+    );
+
+    const result = (await actual.json()) as TrialPaginatedCollection;
```
**suggestion**: I think you can avoid the cast if you use the `toMatchObject` matcher

robbcocco replied:
Updated in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
**issue**: add unit tests to handle the query parameters

robbcocco replied:
Added tests in 8fdbe8398d5cfcd6e97b116f885ad7936c80320f

kin0992 commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +85,48 @@ describe('makeTrialsCosmosContainer', () => {
+      const actual = await makeTrialsCosmosContainer(
+        mockDB as unknown as Database,
+        containerName,
+      ).list({ pageSize: 25, maximumId: undefined, minimumId: undefined })();
```
suggestion: as already said in another comment, the ideal would be:

```suggestion
      ).list({ pageSize: 25 })();
```

robbcocco replied:
Resolved in eff99bbaf41355bc1a98ee131ab9a4d2694136f4

kin0992 reviewed (commented):
I dropped some comments.

**suggestion** for the next activities: try to refine better the activity and create smaller PR. This will help a lot during the code review and speed up the feedback loop

robbcocco committed:
Updated changeset

robbcocco committed:
Update ListTrialOptions type.
Rollback trialId typing.
Add tests for parseQueryParameter.
Update trialId in tests.
Update open api specification.

robbcocco committed:
Fix lint.

kin0992 committed:
[IOPLT-745] Add diagnostic settings to APIM (#253)

robbcocco committed:
Removed ApiTrialSupport from AllowedGroup type.
Updated test.

robbcocco committed:
[IOPLT-772] Add ApiTrialSupport to AllowedGroup type. (#259)

robbcocco committed:
Merge branch 'main' into feature/get_trials

robbcocco committed:
Updated changeset.

robbcocco committed:
Add tests for listtrials query params.

robbcocco committed:
Removed non-relevant changesets.

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/data.ts:
```diff
@@ -123,3 +131,15 @@ export const makeAValidGetTrialRequest = () =>
     },
   });
+
+export const makeAValidListTrialRequest = (params?: {
+  readonly pageSize?: string;
+  readonly maximumId?: string;
+  readonly minimumId?: string;
+}) =>
```
we do not have a type that includes all three parameters, nor a type generated from a schema (we declared the schema in the parameters themselves).

we could use the `ItemId` type for `maximumId` and `minimumId`, but since we're mocking the params sent with the request we can only use string or string-like types (`NumberFromString` in `pageSize` would not work, or would require to stringify the value).

a better solution i think would be to create a dto with this same structure, if necessary

robbcocco committed:
Update changesets.

robbcocco committed:
Update ListTrialOptionsCodec.
Update list trial tests.

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,76 @@
+
+type Env = Pick<SystemEnv, 'listTrials'>;
+
+export type QueryPageSize = t.TypeOf<typeof QueryPageSizeBase>;
+const QueryPageSizeBase = t.union([
+  WithinRangeInteger<1, 100, IWithinRangeIntegerTag<1, 100>>(1, 100),
+  t.literal(100),
+]);
```
this type was generated through `yarn generate` starting from the schema for `queryPageSize` and is used to validate the incoming pageSize parameter.
we put an upper limit of 100 to avoid queries of 100+ trials from the api. it defaults to 25.

lmk if we should increase or lower the limit!

kin0992 replied:
you are right, i forgot the values inside the OpenAPI. 😄

robbcocco committed:
Update makeListTrialsHandler tests.

robbcocco commented on .changeset/grumpy-singers-fry.md:
```diff
@@ -0,0 +1,6 @@
+---
+"functions-subscription": minor
+"@infra/resources": minor
```
Resolved in eb231c5931bdcc59adfb28aca7e7996a0b4bb5cd

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,214 @@
+      previousId: anotherSlimTrial.id,
+      nextId: aSlimTrial.id,
+    });
+  });
```
Resolved in 5adc7847d849a29297b8899a8207645046cf50df

robbcocco commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +85,105 @@ describe('makeTrialsCosmosContainer', () => {
+      const anotherTrial = {
+        ...aTrial,
+        id: 'anotherTrialId' as TrialId,
+        name: 'anotherTrialName' as NonEmptyString,
```
Resolved in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

robbcocco commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +85,105 @@ describe('makeTrialsCosmosContainer', () => {
+
+      expect(actual).toStrictEqual(E.right(trials));
+      expect(mockDB.container('').items.query).toHaveBeenCalledTimes(1);
+    });
```
Resolved in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

robbcocco commented on apps/functions-subscription/src/domain/trial.ts:
```diff
@@ -21,6 +21,17 @@ export const TrialIdCodec = t.brand(
+  }),
+  t.partial({
+    maximumId: t.union([TrialIdCodec, t.undefined]),
+    minimumId: t.union([TrialIdCodec, t.undefined]),
```
Updated in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

robbcocco commented on apps/functions-subscription/src/domain/__tests__/trial.test.ts:
```diff
@@ -23,6 +24,81 @@ describe('insertTrial', () => {
+    const expected = E.right([anotherTrial1, anotherTrial2]);
+
+    expect(actual).toMatchObject(expected);
+  });
```
Resolved in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/process-trial-changes.test.ts:
Resolved in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

robbcocco commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +85,105 @@ describe('makeTrialsCosmosContainer', () => {
+
+      expect(actual).toStrictEqual(E.right(trials));
+      expect(mockDB.container('').items.query).toHaveBeenCalledTimes(1);
+    });
```
Resolved in 2482a6e14320c9c3694301053c8e718cd2f7c0d9

kin0992 commented on .changeset/eight-dots-eat.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@infra/resources": minor
```
**question**: minor or patch?

robbcocco replied:
Updated to patch in acbf3bec67503f6c37e36a4fa82d031c2b8e84ec

kin0992 commented on .changeset/grumpy-singers-fry.md:
```diff
@@ -0,0 +1,5 @@
+"functions-subscription": minor
+---
+
+[IOPLT-746] Add listTrial to CosmosDB and http layer
```
**suggestion**

```suggestion
[IOPLT-746] Create endpoint to get the trial list
```

what do you think?

robbcocco replied:
Updated in acbf3bec67503f6c37e36a4fa82d031c2b8e84ec

kin0992 commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +84,82 @@ describe('makeTrialsCosmosContainer', () => {
+      ).list({ pageSize: 25 })();
+
+      expect(actual).toStrictEqual(E.right(trials));
+      expect(mockDB.container('').items.query).toHaveBeenCalledTimes(1);
+    });
```
**suggestion**: here, and in the other unit tests, I'd check the query as well.

you can use `toBeCalledWith` or `nthCalledWith`

robbcocco replied:
Added in 3dc0bb86816f2a2f2f3fbcf3807f68fc3e45837a

kin0992 commented on apps/functions-subscription/src/adapters/azure/cosmosdb/__tests__/trial.test.ts:
```diff
@@ -84,6 +84,82 @@ describe('makeTrialsCosmosContainer', () => {
+      expect(mockDB.container('').items.query).toHaveBeenCalledTimes(1);
+    });
+
+    it('should return list of trials with id greater than "a"', async () => {
```
**suggestion**: following a previous suggestion, we can rename the test something like

`should make the query with pageSize (or the name used in the query) and minimumId)`

what do you think?

If you agree, I'd be consistent with other unit tests as well

robbcocco replied:
Updated in 3dc0bb86816f2a2f2f3fbcf3807f68fc3e45837a

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,124 @@
+import { NonEmptyString } from '@pagopa/ts-commons/lib/strings';
+
+describe('makeListTrialsHandler', () => {
+  const aSlimTrial = {
+    id: aTrial.id,
+    name: aTrial.name,
+    state: aTrial.state,
+    description: aTrial.description,
+  } as Trial;
+
+  const anotherSlimTrial = {
+    ...aSlimTrial,
+    id: 'anotherTrialId' as TrialId,
+    name: 'anotherTrialName' as NonEmptyString,
+  } as Trial;
```
**issue**: slim trial is a concept introduced into the HTTP layer (coming from OpenAPI).

if you cast to `Trial`, then I assume the objects are not `SlimTrial`.

why not using the `aTrial` object already available into the `data.ts` file?

robbcocco replied:
when using `aTrial`, the return object is different from the former as it get stripped of the author property, making the test fail (`AssertionError: expected { items: [ { …(4) }, { …(4) } ], …(2) } to match object { items: [ { …(5) }, { …(5) } ], …(2) }`).
i think it is correct to expect a TrialSlim object, maybe a solution would be to update `toTrialListAPI` to use a `toSlimTrialAPI` instead of `toTrialAPI`?

robbcocco replied:
reading the following comment i think i get what you meant now. i'll update it following your suggestion.

robbcocco replied:
Updated in 3dc0bb86816f2a2f2f3fbcf3807f68fc3e45837a

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,124 @@
+  it('should return 200 when trials exists', async () => {
+    const env = makeTestSystemEnv();
+
+    env.listTrials.mockReturnValueOnce(
+      TE.right([aSlimTrial, anotherSlimTrial]),
+    );
```
**issue**: following a previous comment, the `listTrials` function returns a list of `Trial`, but the name are different

I'd expect to return something like `aTrial, anotherTrial` of type `Trial` here, and them, when checking the returned value after the handler call, I'd check that the `items` property contains the slim trials.

This means that you can return both a `CREATING` and `CREATED` trials when you invoke the `env.listTrials` method, but the results should contain only the attributes defined into the OpenAPI spec.

robbcocco replied:
Updated in 3dc0bb86816f2a2f2f3fbcf3807f68fc3e45837a

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,124 @@
+
+    const actual = await makeListTrialsHandler(env)(
+      makeAValidListTrialRequest({
+        pageSize: '2',
```
**question**: what would you expect if you pass a string which is not a number?

what if you pass `aString` as value of `pageSize`?

robbcocco replied:
Added test case in 5e8673cfe1b62fcf3bf4f3b8dcdf31fb5542525d

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/codec.ts:
```diff
@@ -56,3 +57,13 @@ export const toTrialAPI = (trial: Trial): TrialAPI => {
+  trials: readonly Trial[],
+): TrialPaginatedCollection => {
+  return {
+    items: trials.map(toTrialAPI),
```
**issue(blocking)**: I think this is not coherent with the OpenAPI

I suggest you to try writing an appropriate test first, to validate what I'm saying, but I think that the `toTrialListAPI` would return every properties of the `Trial` type, meanwhile the OpenAPI specification define that the `TrialSlim` object, which is the value of the elements within the `items` here, should contain only few of them

robbcocco committed:
Update changesets.

robbcocco committed:
Update toTrialListAPI codec.
Update trial list tests.

robbcocco committed:
Add test for error 400 on list trials.

robbcocco committed:
Update QueryPageSize codec.
Update list trials tests.

robbcocco committed:
Update QueryPageSize codec.
Update list trials http tests.

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,149 @@
+import { NonEmptyString } from '@pagopa/ts-commons/lib/strings';
+
+describe('makeListTrialsHandler', () => {
+  const anotherTrial = {
+    ...aTrial,
+    id: 'anotherTrialId' as TrialId,
+    name: 'anotherTrialName' as NonEmptyString,
+  } as Trial;
```
**suggestion**: cast here is not necessary

you can remove the cast, and also the name (it's a different trial since you are updating the id).

kin0992 replied:
cast is redundant somewhere else in this test

robbcocco replied:
Removed casts in ce70af07d12fba443dabfbc5aa2b2a899a27949b

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,149 @@
+
+    expect(env.listTrials).nthCalledWith(1, {
+      pageSize: 25,
+      maximumId: undefined,
+      minimumId: undefined,
```
**suggestion**: i think you can omit those params

```suggestion
```

robbcocco replied:
Removed in ce70af07d12fba443dabfbc5aa2b2a899a27949b

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/codec.ts:
```diff
@@ -56,3 +58,23 @@ export const toTrialAPI = (trial: Trial): TrialAPI => {
     : trialAPI;
 };
+
+export const toSlimTrialAPI = (trial: Trial): TrialSlim => {
```
**suggestion**: what do you think?

just to be coherent with other naming
```suggestion
export const toTrialSlimAPI = (trial: Trial): TrialSlim => {
```

robbcocco replied:
Renamed in ce70af07d12fba443dabfbc5aa2b2a899a27949b

kin0992 reviewed (commented):
I left some comments, nice job so far.

I will complete the review tomorrow 🍻

robbcocco committed:
Updated tests description and data.

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,144 @@
+  const anotherTrial = {
+    ...aTrial,
+    id: 'anotherTrialId' as TrialId,
+  } as Trial;
```
**suggestion**: same as previous commend, unnecessary cast

robbcocco replied:
Removed in 135e4c218e4d43aeae3ea6cb5dff5f2bf85a5fe9

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,144 @@
+  it('should return 200 when trials exists', async () => {
+    const env = makeTestSystemEnv();
+
+    const aSlimTrial = {
+      id: aTrial.id,
+      name: aTrial.name,
+      state: aTrial.state,
+      description: aTrial.description,
+    } as Trial;
```
**issue**: cast.

I think I already comment here before, the `aSlimTrial` type should not be `Trial`.
Anyway, you can omit the type, just make sure the `aSlimTrial` has the property you expect

robbcocco replied:
Resolved in cb9047422e69c939846d97666d4205a5bd3635c8

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/list-trials.test.ts:
```diff
@@ -0,0 +1,144 @@
+      description: aTrial.description,
+    } as Trial;
+
+    const anotherSlimTrial = {
+      ...aSlimTrial,
+      id: 'anotherTrialId' as TrialId,
+    } as Trial;
```
**issue**: same here

robbcocco replied:
Resolved in 135e4c218e4d43aeae3ea6cb5dff5f2bf85a5fe9

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,70 @@
+
+type Env = Pick<SystemEnv, 'listTrials'>;
+
+export const QueryPageSize = t.union([
+  t.undefined,
+  IntegerFromString.pipe(WithinRangeInteger(1, 100)),
+]);
```
**suggestion**: add a short comment to explain those numbers

robbcocco replied:
Added comments in 135e4c218e4d43aeae3ea6cb5dff5f2bf85a5fe9

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,70 @@
+    RTE.apSW(
+      'pageSize',
+      pipe(
+        parseQueryParameter(QueryPageSize, 'pageSize')(req),
+        E.map(O.fromNullable),
+        E.map(O.getOrElseW(() => 25)),
+        RTE.fromEither,
```
**suggestion**: add a short comment to explain this snippet

robbcocco replied:
Added comment in 135e4c218e4d43aeae3ea6cb5dff5f2bf85a5fe9

kin0992 reviewed (commented):
Nice! I dropped some comments!

I'm also waiting the review from @PascalDR for a double check, so we have more eyes on this PR. 👀

kin0992 commented:
@robbcocco before asking the next review (by clicking the icon on top of the page), make sure the branch is aligned with `main`

robbcocco committed:
Merge branch 'main' into feature/get_trials

robbcocco committed:
Add comments.
Update test.

robbcocco commented on apps/functions-subscription/src/adapters/azure/functions/__tests__/data.ts:
```diff
@@ -123,3 +131,11 @@ export const makeAValidGetTrialRequest = () =>
+    url: 'https://function/trials',
+    method: 'GET',
+    headers: supportHttpRequestHeaders,
+    query: query,
```
Updated in ce70af07d12fba443dabfbc5aa2b2a899a27949b

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,72 @@
+
+export const QueryPageSize = t.union([
+  t.undefined,
+  // cast the value to an integer and ensure it's within the 1 to 100 range (1 <= value < 101)
```
```suggestion
  // validates the pageSize value according to the OpenAPI specification
```

kin0992 commented on apps/functions-subscription/src/adapters/azure/functions/list-trials.ts:
```diff
@@ -0,0 +1,72 @@
+      'pageSize',
+      pipe(
+        parseQueryParameter(QueryPageSize, 'pageSize')(req),
+        // set a default pageSize if the value received from the request is null or undefined
```
```suggestion
        // set a default pageSize if the value received from the request is null or undefined (default value specified in OpenAPI specification)
```

robbcocco replied:
Updated in cb9047422e69c939846d97666d4205a5bd3635c8

robbcocco committed:
Removed unnecessary cast.

robbcocco committed:
Updated list trials comments.
Updated list trials test data.

robbcocco committed:
Fixed import on list trials test.

kin0992 reviewed (approved):
LGTM! Nice job @robbcocco!
